Add basic AppVeyor tests for Windows, portable stdin/stdout

charlesdaniels opened the pull request:
As part of this, `--filename` is deprecated, and the `filename` field is removed from the spec.

This touches lots of pieces of code I didn't write, so @jyn514 needs to check I haven't mangled his code too terribly, or otherwise done something overly stupid.

charlesdaniels committed:
Basic AppVeyor config and smoke tests

* This resolves #35 - the smoke tests are implemented, although the one
to tests piping input/output won't pass until #28 is resolved.

charlesdaniels committed:
Patched BitShuffle to handle stdin/stdout on win32

* set_defaults now ensures that args.input and args.output are file
  handles, now nothing in BitShuffle needs to care about paths, just
  file handles.

* This resolves #28

charlesdaniels commented:
Crap. I've hit a difference between python 2 and python 3 - the `file` class is now `TextIOWrapper` in python 3.

charlesdaniels committed:
Fixed forward-incompatible use of 'if x is file'

charlesdaniels committed:
Added AppVeyor status badge to README.rst

charlesdaniels committed:
encode_file now explicitly casts data to bytes

Note that even though all of the file handles we open() are explicitly
done in binary mode, stdin / stdout are utf8 text in Python 3.X. This
commit adds code to use the Python 3.x method (stdin.buffer.read...),
but fails over to the Python 2.X method silently if it dosen't work.

charlesdaniels commented:
Finally fixed all the version & platform specific nonsense. Can @jyn514  make sure I haven't broken anything important and merge if so?

jyn514 commented on README.rst:
```diff
@@ -90,5 +96,5 @@ specified encoding format.
 BitShuffle source code and this README::
 
-        ((<<This is a BitShuffle encoded file, download BitShuffle from https://github.com/charlesdaniels/bitshuffle|1|base64|bz2|0|1|bitshuffle.tar|874c105c72b670fc3b659fe1808a09eae31c7417|QlpoNTFBWSZTWXYWafkABaz/kt5WZaF7//////f//v/v//8EAAAAgAhgD98nsb3gKdHO2S5sNiqpZAVo5jOgBooAAF1laAAxJqZCCZT9DTUyap5iUep4o2TSekwT2VBoPUxNNB7VAA0aNpBpoIEyBTTBNVPRqDRoeo2kzUeSNqADQAGjQANAA0AghSaZMg0GaamQ00DEeoMEyADQA ...
+        ((<<This is a BitShuffle encoded file, download BitShuffle from https://github.com/charlesdaniels/bitshuffle|1|base64|bz2|0|1|874c105c72b670fc3b659fe1808a09eae31c7417|QlpoNTFBWSZTWXYWafkABaz/kt5WZaF7//////f//v/v//8EAAAAgAhgD98nsb3gKdHO2S5sNiqpZAVo5jOgBooAAF1laAAxJqZCCZT9DTUyap5iUep4o2TSekwT2VBoPUxNNB7VAA0aNpBpoIEyBTTBNVPRqDRoeo2kzUeSNqADQAGjQANAA0AghSaZMg0GaamQ00DEeoMEyADQAA00DJoCKepAAAAA ...
```
Getting an error when I try to decode either of these in bitshuffle. Not sure what's going on. Can you see if you can replicate this? The error is also present in the master branch, so it's not a result of this pull request.

charlesdaniels replied:
fixed in ea0db94

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -361,16 +365,30 @@ def infer_mode(args):
+    if type(args.input) is str:
+        try:
+            args.input = open(args.input, 'rb')
+        except Exception as e:
```
Make this 'IOException as e'

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -231,21 +229,27 @@ def main():
-        if is_tmp:
-            os.remove(infile)
+        if is_tmp and tmpfile is not None:
+            os.remove(tmpfile)
```
`if is_tmp and tmpfile`

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -231,21 +229,27 @@ def main():
 
+    args.input.close()
+    args.output.close()
+
```
Perhaps this should be done in `decode()`

jyn514 commented on scripts/test_argument_options.sh:
```diff
@@ -74,13 +74,12 @@ expect_usage_error -t gmander
 
-    printf "\t'Error: Input file not found' \
-             does not match logfile\n\n"
+    echo "Expected error message, but got: $(cat $LOG_FILE)"
```
This appears to be a duplicate of the next line

charlesdaniels replied:
fixed in 017105a

jyn514 commented on scripts/test_argument_options.sh:
```diff
@@ -74,13 +74,12 @@ expect_usage_error -t gmander
 printf "When given bad input, prints file not found... "
 $BITSHUFFLE --input /nonexistent/nope > "$LOG_FILE" 2>&1
-if [ "`cat $LOG_FILE`" = "Error: Input file not found" ]; then
+if grep -i 'could not open' < "$LOG_FILE" > /dev/null 2>&1 ; then
```
Are we certain `grep -i` is portable? It works fine on my machine but I'd like to be sure.

charlesdaniels replied:
I have never seen a machine that implements `grep` but not `grep -i`. IIRC it's a pretty old feature for grep. I know for sure that it is present on macOS, Linux, and FreeBSD.

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -361,16 +365,30 @@ def infer_mode(args):
+    if type(args.output) is str:
+        try:
+            args.output = open(args.output, 'wb')
+        except Exception as e:
```
IOException

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -361,16 +365,30 @@ def infer_mode(args):
 def set_defaults(args):
 
-    defaults = {'input': '/dev/stdin', 'output': '/dev/stdout',
+    defaults = {'input': stdin, 'output': stdout,
```
Well done! Didn't even consider that we could do this.

charlesdaniels replied:
This is one of my favorite python-isms. It's also nice because it's entirely out-of-band from the text console.

jyn514 reviewed (changes requested):
Overall, good job. A few minor nitpicks. Most concerningly, I am unable to decode the README in either this branch or the master branch. Can you try to replicate this?

If possible I would like to split this into several different requests, but I understand that might be difficult to do.

jyn514 committed:
Swallow fewer exceptions

Changed several `except Exception`s to `except IOError`
Check `isinstance(file_type)` instead of `is not str`

jyn514 committed:
Don't assume file is text-based

jyn514 committed:
remove debugging messages

charlesdaniels commented:
re: 771b968

Did not realize that `IoBase` existed, very nice :)

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -380,6 +380,7 @@ def set_defaults(args):
     # open args.input and args.output so they are file handles
     if not isinstance(args.input, file_type):
+        stderr.write("opening input")
         try:
```
While I should probably have remembered to remove this, are we certain we want debug messages to fail our build tests? See https://ci.appveyor.com/project/charlesdaniels/bitshuffle/build/38 for context.

charlesdaniels committed:
Removed redundant output from test_argument_options

charlesdaniels committed:
removed example from README until format is stable
